### PR TITLE
Express refactor: Static file routes and templates

### LINKED ISSUE #1292: Allow direct access to files under a path
I am trying to open an HTML file on my PC. code-server is being hosted on a spare PC. I can open and view the file just file in the cloud IDE, but I'm trying to view the HTML file as it would be shown on a website. I've tried navigating to "MYIPADDRESS:8080/home/MYUSERNAME/Projects/alarm-clock/index.html", but that doesn't work. It's probably a very obvious answer, but I haven't been able to find anything about it here.

### PR DESCRIPTION
**Pull request by TeffenEllis**
This PR continues the [Expressification](https://github.com/cdr/code-server/tree/code-asher/ch1385) of code-server.

## Progress

- Refactors asset path helpers and removes `csStaticPath` concept
- Fixed issue where JSON entities may be unescaped
- Updated web manifest schema, route handler
- Fixed issue where body background color is often out of sync on startup and after theme change
- Added types to `vscode` route. 
- Adds `/local` route for authenticated static file serving

Closes #1292

### CODE REVIEW

File: src/node/routes/index.ts [hunk 1]
```diff
@@ -101,20 +105,19 @@ export const register = async (app: Express, server: http.Server, args: Defaulte
...
-        replaceTemplates(req, content)
-          .replace(/{{ERROR_TITLE}}/g, status)
-          .replace(/{{ERROR_HEADER}}/g, status)
-          .replace(/{{ERROR_BODY}}/g, err.message),
-      )
+      res.status(status).render("error", {
+        HOME_PATH: (typeof req.query.to === "string" && req.query.to) || "/",
+        ERROR_TITLE: status,
+        ERROR_HEADER: status,
+        ERROR_BODY: err.message,
```
Comment (code-asher) [inline src/node/routes/index.ts:119]: I really like how clean these template vars became.

File: src/node/routes/manifest.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import { Router } from "express"
+import { normalize } from "../../common/util"
+import { CommonTemplateVars, commonTemplateVars } from "../http"
+
+export const router = Router()
+
+const iconSizes = [96, 128, 192, 256, 384, 512]
```
Comment (code-asher) [inline src/node/routes/manifest.ts:7]: Dope that we don't have to manually add each one now.

File: src/browser/views/error.handlebars [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    />
+    <meta
+      http-equiv="Content-Security-Policy"
+      content="style-src 'self'; manifest-src 'self'; img-src 'self' data:; font-src 'self' data:;"
+    />
+    <title>{{ERROR_TITLE}} — Code Server</title>
+    <link rel="icon" href="{{assetPath csStaticBase '/media/favicon.ico'}}" type="image/x-icon" />
+    <link rel="manifest" href="{{assetPath base '/code-server.webmanifest'}}" crossorigin="use-credentials" />
+    <link rel="apple-touch-icon" href="{{assetPath csStaticBase '/media/pwa-icon-384.png'}}" />
+    <link href="{{assetPath csStaticBase '/register.css'}}" rel="stylesheet" />
```
Comment (code-asher) [inline src/browser/views/error.handlebars:17]:
```suggestion
    <link href="{{assetPath base '/register.css'}}" rel="stylesheet" />
```

Or `<link href="{{assetPath base '/dist/register.css'}}" rel="stylesheet" />` if we end up moving the `dist` endpoint.

File: src/node/app.ts [hunk 4]
```diff
@@ -1,16 +1,43 @@
...
+  app.engine(
+    "handlebars",
+    exphbs({
+      defaultLayout: "",
+      helpers: {
+        prod: () => prod,
+        assetPath: (base: string, path: string) => normalize(base + path),
+        /**
+         * Converts to JSON string and encodes entities for use in HTML.
+         * @TODO we can likely move JSON attributes to <script type="text/json">
```
Comment (code-asher) [inline src/node/app.ts:31]: :+1: love this idea

**Review by code-asher — CHANGES_REQUESTED** [review]
Handlebars + static combo-wombo! This looks great! ty again for taking care of this.

File: src/node/routes/login.ts [hunk 5]
```diff
@@ -90,6 +82,6 @@ router.post("/", async (req, res) => {
 
     throw new Error("Incorrect password")
   } catch (error) {
-    res.send(await getRoot(req, error))
+    rootHandler(error, req, res, next)
```
[the remark this answers is not in the document]
Comment (TeffenEllis) [reply src/node/routes/login.ts:85]: I believe it does just as a virtue of using the `ErrorRequestHandler` 🙃

[the remark this answers is not in the document]
Comment (code-asher) [reply src/node/routes/login.ts:86]: Ahh oops makes sense.

**Review by code-asher — COMMENTED** [review]
Latest changes look great! There are still some unresolved comments (looks like they're being collapsed and hidden though).

**Review by Qreig — APPROVED** [review]

**Comment by code-asher** [comment]
I haven't been able to re-review this yet. The divergence is growing and I'm not totally sure the best way to proceed but here's my plan:

1. Close to keep the PR queue slim.
2. Rebase on master.
3. Split up the commit a bit.
4. Open a new PR.

Additionally, we had a discussion about Handlebars and we want to try taking a route that lets us have type checking in the template (if we do though it'll basically just be replacing the `.hbs` files, so not a huge departure).

**Comment by nhooyr** [comment]
Deleted this branch for now. I have it locally and saved on my fork in case we need to access these commits again. Just trying to keep the main repositories branches clean.

**Comment by nhooyr** [comment]
For posterity, the link is https://github.com/nhooyr/code-server/tree/z/handlebars-templates

**Pull request closed without merging**